hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java

```diff
@@ -0,0 +1,196 @@
+        this.stop(this.currentInstantTime);
+      }
+      this.currentInstantTime = instantTime;
+      this.numHeartBeatsForCurrentInstantTime = 0;
+      this.shutdownRequested = false;
+    }
+    try {
+      updateHeartbeat(instantTime);
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to generate heartbeat");
+    }
+    this.heartBeatFuture = executorService.submit(() -> {
+      while (!shutdownRequested) {
+        try {
+          Thread.sleep(this.heartbeatIntervalInMillis);
+          updateHeartbeat(instantTime);
+        } catch (InterruptedIOException | InterruptedException ioe) {
+          LOG.warn("Thread controlling heartbeat was interrupted");
+        } catch (IOException ioe) {
+          LOG.error("Unable to create heartbeat file", ioe);
+          throw new RuntimeException(ioe);
+        }
+      }
+    });
+    this.isHearbeatStarted = true;
+    this.isHeatBeatStopped = false;
+  }
+
+  public void stop(String instantTime) throws HoodieException {
+    this.shutdownRequested = true;
+    if (isHearbeatStarted && !isHeatBeatStopped && this.heartBeatFuture != null) {
+      // TODO : Substract the amount of time it takes to execute updateHeartbeat from the sleep time to avoid race conditions
+      boolean isCancelled = this.heartBeatFuture.cancel(true);
+      if (!isCancelled) {
+        throw new HoodieException("Could not stop heartbeat client");
+      }
+      this.isHeatBeatStopped = true;
+      LOG.info("Stopping heartbeat for instant " + instantTime);
+    }
+  }
+
+  public void stop() throws HoodieException {
+    if (this.currentInstantTime != null) {
+      stop(this.currentInstantTime);
+    }
+  }
+
+  public boolean delete(String instantTime) throws IOException {
+    // TODO ensure that we are not deleting the current instant heartbeat, this might need storing the instant time
+    // and implementing another heartbeat utils class for everything else
+    if (this.currentInstantTime == instantTime && !this.isHeatBeatStopped) {
+      LOG.error("Cannot delete a currently running heartbeat, stop it first");
+      return false;
+    } else {
+      boolean deleted = this.fs.delete(new Path(heartBeatFolderPath + File.separator + instantTime), false);
+      if (!deleted) {
+        LOG.error("Failed to delete heartbeat for instant " + instantTime);
+      }
+      return deleted;
+    }
+  }
+
+  public static Long getLastHeartbeat(FileSystem fs, String basePath, String instantTime) throws IOException {
+    Path heartBeatFilePath = new Path(HoodieTableMetaClient.getHeartbeatFolderPath(basePath) + File.separator + instantTime);
+    if (fs.exists(heartBeatFilePath)) {
+      return fs.getFileStatus(heartBeatFilePath).getModificationTime();
+    } else {
+      // NOTE : This can happen when a writer is upgraded to use lazy cleaning and the last write had failed
+      return 0L;
+    }
+  }
+
+  public boolean checkIfConcurrentWriterRunning(String instantTime) throws IOException {
```

Comment:
this is leaking upper level context (multi writing) to a lower level class. We should just name this something around heartbeats. say `isHeartbeatExpired()`